csharp/src/Ice/VectoredBuffer.cs

```diff
@@ -0,0 +1,549 @@
+                Size = 0;
+                Capacity = MinSegmentSize;
+            }
+            else
+            {
+                Segments.Add(new ArraySegment<byte>(initialData));
+                Size = initialData.Length;
+                Capacity = initialData.Length;
+            }
+            _currentSegment = Segments[0];
+            _currentSegmentOffset = 0;
+        }
+
+        /// <summary>
+        /// Create a buffer from a list of byte arrays, it array is added as a segment to the buffer,
+        /// the Size and Capacity are set to the summ of the arrays lengths.</summary>
+        /// <param name="segments"></param>
+        public VectoredBuffer(byte[][] segments)
+        {
+            Segments = new List<ArraySegment<byte>>();
+            _position = new BufferPosition(0, 0);
+            foreach (byte[] segment in segments)
+            {
+                Segments.Add(segment);
+                Size += segment.Length;
+            }
+            Capacity = Size;
+            _currentSegment = Segments[0];
+            _currentSegmentOffset = 0;
+        }
+
+        /// <summary>Add additional data segments to the stream, the segments are append to the end
+        /// of the segments list, Prepare must be call before calling this to ensure there
+        /// is not gaps when adding segments.</summary>
+        /// <param name="segments">The list of segments to add to the buffer.</param>
+        public void Add(IList<ArraySegment<byte>> segments)
+        {
+            foreach (ArraySegment<byte> segment in segments)
+            {
+                Segments.Add(segment);
+                Size += segment.Count;
+            }
+            Capacity = Size;
+        }
+
+        public void Add(ArraySegment<byte> segment)
+        {
+            Segments.Add(segment);
+            Size += segment.Count;
+            Capacity = Size;
+        }
+
+        /// <summary>Advanced the buffer position the given number of bytes.</summary>
+        /// <param name="size">The number of bytes to advance the buffer position.</param>
+        public void Advance(int size)
+        {
+            Debug.Assert(size <= Remaining);
+            while (size > 0)
+            {
+                ArraySegment<byte> segment = Segments[_position.Segment];
+                int remaining = segment.Count - _position.Offset;
+                if (remaining >= size)
+                {
+                    _position.Offset += size;
+                    size -= remaining;
+                }
+                else
+                {
+                    size -= remaining;
+                    _currentSegmentOffset += segment.Count;
+                    _position.Offset = 0;
+                    _position.Segment++;
+                    _currentSegment = Segments[_position.Segment];
+                }
+            }
+        }
+
+        /// <summary>Release all the data hold by this buffer</summary>
+        public void Clear()
```

Comment:
Why is this method necessary?

Reply:
This is called in a few places when the data held by the buffer is no longer required

https://github.com/pepone/ice/blob/vectored-io/csharp/src/Ice/StreamSocket.cs#L132

https://github.com/pepone/ice/blob/vectored-io/csharp/src/Ice/WSTransceiver.cs#L304

That call to Reset I think should also be calling Clear instead.

https://github.com/pepone/ice/blob/vectored-io/csharp/src/Ice/Connection.cs#L1485